File: java/lld/vending-machine/VendingMachine.java
import java.util.Map;

// The CONTEXT — one 'synchronized' lock point makes the machine thread-safe.
public class VendingMachine {
    private State state = new IdleState();
    int balance = 0;
    String selected = null;
    final Map<String, Product> products;

    public VendingMachine(Map<String, Product> products) {
        this.products = products;
    }

    void setState(State s) { this.state = s; }
    void returnChange(int cents) { System.out.println("return " + cents + "c"); }

    public synchronized void insertCoin(int cents)     { state.insertCoin(this, cents); }
    public synchronized void selectProduct(String id)  { state.selectProduct(this, id); }
    public synchronized void dispense()                { state.dispense(this); }
    public synchronized void refund()                  { state.refund(this); }

    static class Product {
        int price, stock;

        Product(int p, int s) { price = p; stock = s; }
    }
}

File: python/lld/vending-machine/vending_machine.py
from threading import Lock
from idle_state import IdleState

class Product:
    def __init__(self, price, stock):
        self.price = price
        self.stock = stock

class VendingMachine:
    """The CONTEXT — one lock makes the whole machine thread-safe."""

    def __init__(self, products):
        self._state = IdleState()
        self._lock = Lock()
        self.balance = 0
        self.selected = None
        self.products = products

    def set_state(self, s): self._state = s
    def return_change(self, cents): print(f"return {cents}c")

    def insert_coin(self, cents):
        with self._lock: self._state.insert_coin(self, cents)

    def select_product(self, pid):
        with self._lock: self._state.select_product(self, pid)

    def dispense(self):
        with self._lock: self._state.dispense(self)

    def refund(self):
        with self._lock: self._state.refund(self)